File: internal/util/spec.go
package util

import (
	"errors"
	"os"
	"path/filepath"
	"strings"
)

const DefaultSpecContent = `# Goal

<describe the goal>

# Constraints / nuances

- <constraints>

# Acceptance tests

- <acceptance tests>
`

func DefaultSpecPath(workspacePath, name string) (string, error) {
	if strings.TrimSpace(workspacePath) == "" {
		return "", errors.New("workspace path is empty")
	}
	if strings.TrimSpace(name) == "" {
		return "", errors.New("spec name is empty")
	}
	absWorkspace, err := filepath.Abs(workspacePath)
	if err != nil {
		return "", err
	}
	return filepath.Join(absWorkspace, "specs", name, "spec.md"), nil
}

func EnsureSpecFile(path string) (bool, error) {
	if strings.TrimSpace(path) == "" {
		return false, errors.New("spec path is empty")
	}
	if _, err := os.Stat(path); err == nil {
		return false, nil
	} else if !errors.Is(err, os.ErrNotExist) {
		return false, err
	}
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return false, err
	}
	content := DefaultSpecContent
	if !strings.HasSuffix(content, "\n") {
		content += "\n"
	}
	if err := os.WriteFile(path, []byte(content), 0o644); err != nil {
		return false, err
	}
	return true, nil
}

File: internal/util/spec.ts
import path from "node:path";
import { stat, mkdir, writeFile } from "node:fs/promises";

export const defaultSpecContent = `# Goal

<describe the goal>

# Constraints / nuances

- <constraints>

# Acceptance tests

- <acceptance tests>
`;

export async function defaultSpecPath(workspacePath: string, name: string): Promise<string> {
  if (!workspacePath.trim()) throw new Error("workspace path is empty");
  if (!name.trim()) throw new Error("spec name is empty");
  const absWorkspace = path.resolve(workspacePath);
  return path.join(absWorkspace, "specs", name, "spec.md");
}

export async function ensureSpecFile(specPath: string): Promise<boolean> {
  if (!specPath.trim()) throw new Error("spec path is empty");
  try {
    await stat(specPath);
    return false;
  } catch {
    // continue
  }
  await mkdir(path.dirname(specPath), { recursive: true, mode: 0o755 });
  const content = defaultSpecContent.endsWith("\n") ? defaultSpecContent : `${defaultSpecContent}\n`;
  await writeFile(specPath, content, { mode: 0o644 });
  return true;
}
